### PR TITLE
avoid panic in discover when no st link is connected and add examples

### PR DESCRIPTION
**Pull request by ckrenslehner**
Hi!

I added some small changes to avoid calling from_raw_parts when there is null ptr.
How is your general experience with discovering ST-Links? I tried the `discover` example I added multiple times in a row. Sometimes the ST-Link is detected, sometimes not. It is unstable on my machine. I have version 2.14 of the CubeProgrammer.

Which version did you use for testing?

Greetings, Christian

### CODE REVIEW

**Comment by wervin** [comment]
Hi!

Thanks for your work! I'll check it out as soon as possible.

Regarding the version, I'm using v2.14 on Ubuntu. I'm not exactly sure, but I did experience the same issue and haven't looked into it yet.

Some functions, like Read Core Registers, are not detailed in the API header but are somehow available in the DLL.

Bye, Vincent

**Comment by ckrenslehner** [comment]
Hi! 

The issue is v2.14. I updated to v2.18 and the issue does not occur anymore. :-)

I have started with my [own wrapper](https://github.com/ckrenslehner/stm32cubeprogrammer-rs), inspired by your project.
Maybe you want to check it out, give some feedback, or join forces. :-)

Greetings, Christian

**Review by wervin — CHANGES_REQUESTED** [review]
Hey, awesome work on this PR! 🚀 I really like all the changes you’ve made—great stuff. The only thing I’m not thrilled about is the removal of the read and write register interface. Also, I noticed you’ve started your own repo—nice! I’ll make sure to reference it in the README. 😄

**Comment by ckrenslehner** [comment]
The removal of these functions was a kind of an accident, so please just ignore it 😄 how did you find out about these functions?

Yes I started my own repo, because I needed support for stm32wb and was suddenly very interested in this task 😄 Also, I don't really like the official CLI of the cube programmer. That's why I want to make one on my own. :-)
Exposing the dll logs via a trate offers the possibility to integrate e.g. a progress bar in a cli or a tui. A ratatui based tui for programming would be a fun project for sure 😄

But as mentioned, maybe we can combine the projects and publish them under a single crate. :-)

Greetings

**Comment by wervin** [comment]
You can use tools like Ghidra or Radare, for example, to reverse-engineer the library and uncover undocumented APIs, that’s exactly what I did. It’s a bit odd that we need to resort to this, but it is what it is... 

As mentioned in the README, I don’t have much time to improve the project right now, as I’m focusing on learning KiCad and developing my own boards. It might be better if you take the lead on this :rocket: 

However, I’m happy to test your crate on my machine if needed :smile_cat: Also, if you have any questions about the FUS of the STM32WB, I’m happy to provide support. 

Additionally, I also have a Python package similar to this Rust package, which includes both FUS and DFU capabilities.

**Comment by ckrenslehner** [comment]
I added the register functions again. 

Thanks for the tip of using Ghidra or Radare. I will take a look at them. :-)
I appreciate your offer for testing. As soon as I consider it as "stable" I will tell you, thanks!

Indeed, I have a question to the FUS. I cannot find how to read the version of the currently installed BLE stack in the API docs / examples. Do you know how to get this info by chance?
The CubeProgrammer GUI shows the information, so there certainly is a way to get it via the API

**Comment by ckrenslehner** [comment]
Btw. Phil's Lab has a lot of great vidoes about PCB design principles. In some of them he uses KiCAD
https://www.youtube.com/watch?v=nkHFoxe0mrU&ab_channel=Phil%E2%80%99sLab

File: src/lib.rs [hunk 1]
```diff
@@ -246,15 +247,16 @@ pub struct VTable {
...
     disconnect: libloading::os::unix::Symbol<Disconnect>,
     reset: libloading::os::unix::Symbol<Reset>,
     mass_erase: libloading::os::unix::Symbol<MassErase>,
     download_file: libloading::os::unix::Symbol<DownloadFile>,
     get_device_general_info: libloading::os::unix::Symbol<GetDeviceGeneralInfo>,
     read_memory: libloading::os::unix::Symbol<ReadMemory>,
     write_memory: libloading::os::unix::Symbol<WriteMemory>,
-    read_core_register: libloading::os::unix::Symbol<ReadCoreRegister>,
-    write_core_register: libloading::os::unix::Symbol<WriteCoreRegister>,
+    read_core_register: libloading::os::windows::Symbol<ReadCoreRegister>,
```
Comment (wervin) [inline src/lib.rs:258]: It should be unix in this case :smile_cat:

**Review by wervin — CHANGES_REQUESTED** [review]
Some typos remain in the Unix VTables :stuck_out_tongue:

**Comment by wervin** [comment]
> Btw. Phil's Lab has a lot of great vidoes about PCB design principles. In some of them he uses KiCAD https://www.youtube.com/watch?v=nkHFoxe0mrU&ab_channel=Phil%E2%80%99sLab

Thank you ! I want to develop a simple board for the SGTL5000 audio codec for my need, it will be useful

**Comment by wervin** [comment]
> I added the register functions again.
> 
> Thanks for the tip of using Ghidra or Radare. I will take a look at them. :-) I appreciate your offer for testing. As soon as I consider it as "stable" I will tell you, thanks!
> 
> Indeed, I have a question to the FUS. I cannot find how to read the version of the currently installed BLE stack in the API docs / examples. Do you know how to get this info by chance? The CubeProgrammer GUI shows the information, so there certainly is a way to get it via the API

Regarding the FUS, you can retrieve the Wireless Stack (WS) version from the device info table located in the shared SRAM2 memory.

Note that depending on whether the FUS or WS is running, the information might not be in the same location. You can refer to the file Middlewares/ST/STM32_WPAN/interface/patterns/ble_thread/shci/shci.c, specifically the function SHCI_GetWirelessFwInfo.

Using STM32CubeProgrammer, if the FUS is running, you can obtain the WS version as follows:

image

**Comment by ckrenslehner** [comment]
Okay thanks! Indeed I already read the stack version in my firmware like that. I just thought there may be a "higher" level way of doing things using the API. But this is fine too. Thank you again.

**Comment by wervin** [comment]
> Okay thanks! Indeed I already read the stack version in my firmware like that. I just thought there may be a "higher" level way of doing things using the API. But this is fine too. Thank you again.

There might be a hidden function in the library, such as readVersion or something similar, but I haven't investigated it yet 😺

File: src/lib.rs (hunk 1 above)
Comment (ckrenslehner) [reply src/lib.rs:258]: Oh sorry I missed that!

**Pull request merged**